### PR TITLE
Destroy model: watch better suited collection for changes.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

When destroying models, we used to watch only applications and machines collections. However, when we were determining if model is ready to be destroyed, we'd check collection that stored model entities references.

This introduced 2 problems:
1. If we were to identify new entities that model should worry about before self-destructing, we would miss them as we were not watching their collections.
2. There was a potential for a race - which was observed fairly reliably - between when a specific entity collection would change and when the reference for that collection would be dealt with.  

This PR changes undertaker behavior to watch for changes on collection that references model entities rather than said entities collections.

We are also now surfacing errors that we've encountered during the attempts to destroy model as well as log them.
Drive-by fixes:

## QA steps

There are several scenarios where destroy-model would chase its own tail.
The most commonly encountered one that I have used:
1. bootstrap;
2. add model;
3. deploy an application with one unit - wait for dust to settle;
4. remove that one unit - wait for dust to settle;
[you would have an empty deployed app]
5. destroy this model.
Model gets destroyed successfully.

Previously, model destruction would get into infinite loop and only eventual kill-controller will get rid of it.
Note that if you are running 'watch juju status' during model destruction, you may not observe failure to destroy.

## Documentation changes

n/a - internal change

## Bug reference

https://bugs.launchpad.net/juju/+bug/1672549
https://bugs.launchpad.net/juju/+bug/1661930

### CODE REVIEW

File: apiserver/undertaker/state.go [hunk 1]
```diff
@@ -31,62 +30,18 @@ type State interface {
...
-	// AllMachines returns all machines in the model ordered by id.
-	AllMachines() ([]Machine, error)
-
-	// AllApplications returns all deployed services in the model.
-	AllApplications() ([]Service, error)
-
 	// ModelConfig retrieves the model configuration.
 	ModelConfig() (*config.Config, error)
+
+	// WatchModelEntitiesReferences gets a watcher capable of monitoring
```
Comment (axw) [inline apiserver/undertaker/state.go]: s/Entities/Entity/

File: apiserver/undertaker/undertaker.go [hunk 2]
```diff
@@ -85,27 +85,18 @@ func (u *UndertakerAPI) RemoveModel() error {
...
-		watchers = append(watchers, machine.Watch())
-	}
-	for _, service := range services {
-		watchers = append(watchers, service.Watch())
-	}
 
-	watch := common.NewMultiNotifyWatcher(watchers...)
+	// (anastasiamac 2017-04-10) I do not know what watcher is best suited here.
+	// Keeping multi-notify watcher for historical reasons for now.
+	watch := common.NewMultiNotifyWatcher(u.st.WatchModelEntitiesReferences(m.UUID()))
```
Comment (axw) [inline apiserver/undertaker/undertaker.go]: Drop the multi and use the entity references watcher directly. There is now and for the foreseeable future should only be one.

File: state/application.go [hunk 3]
```diff
@@ -272,10 +272,6 @@ func (a *Application) removeOps(asserts bson.D) ([]txn.Op, error) {
 			Id:     a.doc.DocID,
 			Assert: asserts,
 			Remove: true,
-		}, {
```
Comment (axw) [inline state/application.go]: please move this to another PR if it is still relevant

File: state/application.go [hunk 4]
```diff
@@ -290,7 +286,6 @@ func (a *Application) removeOps(asserts bson.D) ([]txn.Op, error) {
 		return nil, errors.Trace(err)
 	}
 	ops = append(ops, charmOps...)
-	ops = append(ops, finalAppCharmRemoveOps(name, curl)...)
```
Comment (axw) [inline state/application.go]: and this

File: state/charm_test.go [hunk 5]
```diff
@@ -256,12 +256,10 @@ func (s *CharmSuite) TestDestroyFinalUnitReference(c *gc.C) {
 	app := s.Factory.MakeApplication(c, &factory.ApplicationParams{
 		Charm: s.charm,
 	})
-	unit := s.Factory.MakeUnit(c, &factory.UnitParams{
```
Comment (axw) [inline state/charm_test.go]: and these

File: state/model.go [hunk 6]
```diff
@@ -1040,16 +1046,22 @@ func (m *Model) checkEmpty() error {
 		}
 		return errors.Annotatef(err, "getting entity references for model %s", m.UUID())
 	}
+	// These errors could be potentially swallowed as we re-try to destroy model.
+	// Let's, at least, log them for observations.
```
Comment (jameinel) [inline state/model.go]: observation.

File: worker/undertaker/undertaker.go [hunk 7]
```diff
@@ -172,7 +175,10 @@ func (u *Undertaker) processDyingModel() error {
 				// destroy any remaining environ resources.
 				return nil
 			}
-			// Yes, we ignore the error. See comment above.
+			// Yes, we ignore the error. See comment above. But let's at least
+			// surface it in status.
+			u.setStatus(status.Destroying, fmt.Sprintf("%d attempt to destroy model:  %v", attempt, err))
```
Comment (jameinel) [inline worker/undertaker/undertaker.go]: 2 attempt to destroy model: error
would probably read better as:
"attempt %d to destroy model failed (will retry): %v"

moving the counter helps readability, and some indication as to what we are doing about this failure.
These are going to be shown in "juju status", so making them legible is important.

File: apiserver/undertaker/state.go [hunk 8]
```diff
@@ -31,62 +30,18 @@ type State interface {
...
-	AllMachines() ([]Machine, error)
-
-	// AllApplications returns all deployed services in the model.
-	AllApplications() ([]Service, error)
-
 	// ModelConfig retrieves the model configuration.
 	ModelConfig() (*config.Config, error)
+
+	// WatchModelEntityReferences gets a watcher capable of monitoring
+	// model entities references changes.
```
Comment (axw) [inline apiserver/undertaker/state.go]: s/entities/entity/

File: apiserver/undertaker/undertaker.go [hunk 9]
```diff
@@ -85,28 +85,16 @@ func (u *UndertakerAPI) RemoveModel() error {
...
+func (u *UndertakerAPI) modelEntitiesWatcher() params.NotifyWatchResult {
 	var nothing params.NotifyWatchResult
-	machines, err := u.st.AllMachines()
-	if err != nil {
-		nothing.Error = common.ServerError(err)
-		return nothing
-	}
-	services, err := u.st.AllApplications()
+
+	m, err := u.st.Model()
```
Comment (axw) [inline apiserver/undertaker/undertaker.go]: just call state.ModelUUID instead? no need to hit the db just to get the UUID

File: apiserver/undertaker/state.go (hunk 1 above)
Comment (anastasiamac) [inline apiserver/undertaker/state.go]: Changed :D nice pick-up! I was, of course thinking of model entities as a concept.

File: apiserver/undertaker/undertaker.go (hunk 2 above)
Comment (anastasiamac) [inline apiserver/undertaker/undertaker.go]: Dropped. Thank you for advice!

File: state/application.go (hunk 3 above)
Comment (anastasiamac) [inline state/application.go]: Removed this and the other related bits. Will propose separately :) I do not think that it is particularly hurtful but was very surprising to examine a transaction where several exact same operations appeared...

File: state/model.go (hunk 6 above)
Comment (anastasiamac) [inline state/model.go]: Nice :) With my English, both plural and singular work here but I appreciate your feedback! Changed :)

File: worker/undertaker/undertaker.go (hunk 7 above)
Comment (anastasiamac) [inline worker/undertaker/undertaker.go]: Great idea! Reads better indeed. Thank you :D

File: apiserver/undertaker/undertaker.go (hunk 9 above)
Comment (anastasiamac) [inline apiserver/undertaker/undertaker.go]: oh k.. I was looking at other code in the area. I'll take a look to see what state.ModelUUID actually gives here :)

File: apiserver/undertaker/state.go (hunk 8 above)
Comment (anastasiamac) [inline apiserver/undertaker/state.go]: Of course, I'll change here and in other comments..

File: apiserver/undertaker/undertaker.go (hunk 9 above)
Comment (anastasiamac) [inline apiserver/undertaker/undertaker.go]: done \o/ so much neater!

**Comment by anastasiamac** [comment]
$$merge$$
